src/Abstractions/TaskName.cs

```diff
@@ -8,127 +8,162 @@ namespace Microsoft.DurableTask;
-    /// </summary>
-    /// <value>
-    /// The name of the activity task without the version.
-    /// </value>
-    public string Name { get; }
+  /// <summary>
+  /// Initializes a new instance of the <see cref="TaskName"/> struct.
+  /// </summary>
+  /// <param name="name">The name of the task. Providing <c>null</c> will yield the default struct.</param>
+  /// <param name="version">The version of the task.</param>
+  public TaskName(string name, string version)
+  {
+    if (name is null)
+    {
+      if (version is null)
+      {
+        // Force the default struct when null is passed in.
+        this.Name = null!;
+        this.Version = null!;
+      }
+      else
+      {
+        throw new ArgumentException("name must not be null when version is non-null");
+      }
+    }
+    else
+    {
+      if (version is null)
+      {
+        this.Name = name;
+        this.Version = string.Empty; // fallback to the contructor without version parameter
+      }
+      else
+      {
+        this.Name = name;
+        this.Version = version;
+      }
+    }
+  }
 
-    /// <summary>
-    /// Gets the version of the task.
-    /// </summary>
-    /// <remarks>
-    /// Task versions is currently locked to <see cref="string.Empty" /> as it is not yet integrated into task
-    /// identification. This is being left here as we intend to support it soon.
-    /// </remarks>
-    public string Version { get; }
+  /// <summary>
+  /// Gets the name of the task without the version.
+  /// </summary>
+  /// <value>
+  /// The name of the activity task without the version.
+  /// </value>
+  public string Name { get; }
 
-    /// <summary>
-    /// Implicitly converts a <see cref="TaskName"/> into a <see cref="string"/> of the <see cref="Name"/> property value.
-    /// </summary>
-    /// <param name="value">The <see cref="TaskName"/> to be converted into a string.</param>
-    public static implicit operator string(TaskName value) => value.Name;
+  /// <summary>
+  /// Gets the version of the task.
+  /// </summary>
+  /// <remarks>
+  /// Task versions is currently locked to <see cref="string.Empty" /> as it is not yet integrated into task
+  /// identification. This is being left here as we intend to support it soon.
+  /// </remarks>
+  public string Version { get; }
 
-    /// <summary>
-    /// Implicitly converts a <see cref="string"/> into a <see cref="TaskName"/> value.
-    /// </summary>
-    /// <param name="value">The string to convert into a <see cref="TaskName"/>.</param>
-    public static implicit operator TaskName(string value) => string.IsNullOrEmpty(value) ? default : new(value);
+  /// <summary>
+  /// Implicitly converts a <see cref="TaskName"/> into a <see cref="string"/> of the <see cref="Name"/> property value.
+  /// </summary>
+  /// <param name="value">The <see cref="TaskName"/> to be converted into a string.</param>
+  public static implicit operator string(TaskName value) => value.Name;
```

Comment:
This doesn't call `value.ToString()`? (Shocked I missed that previously). Should definitely fix that.

Reply:
Updated as you suggestion.